File: python/547.py
from typing import List
import collections
import math
from functools import lru_cache
import queue
import bisect

class UF_balanced:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.depth = [1] * n

    def find(self, x):
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        xp, yp = self.find(x), self.find(y)
        if xp == yp:
            return False
        ## union
        if self.depth[xp] < self.depth[yp]:
            t = xp
            xp = yp
            yp = t
        self.parent[xp] = self.parent[yp]
        self.depth[xp] += self.depth[yp]
        return True

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        uf = UF_balanced(n)
        for i in range(n):
            for j in range(n):
                if isConnected[i][j]:
                    uf.union(i, j)

        ret = 0
        for i in range(n):
            if i == uf.parent[i]:
                ret += 1
        return ret



if __name__ == '__main__':
    sol = Solution()
    s = sol.findCircleNum([[1,0,0],[0,1,0],[0,0,1]])
    print(s)
    s = sol.findCircleNum([[0]])
    print(s)
    s = sol.findCircleNum([[1,1,0],[1,1,0],[0,0,1]])
    print(s)

File: cpp/547.cpp
#include <bits/stdc++.h>

using namespace std;

class UF {
public:
    vector<int> parent;
    
    UF(int n) : parent(vector<int>(n)) {
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
    }

    int getParent(int x) {
        if (x != parent[x]) {
            parent[x] = getParent(parent[x]);
        }
        return parent[x];
    }

    void add(int u, int v) {
        int pu, pv;
        pu = getParent(u), pv = getParent(v);
        if (pu != pv) {
            parent[pu] = pv;
        }
    }
};

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n =  isConnected.size();
        UF uf(n);
        for (int i = 0; i < n; ++i) {
            for (int j = i+1; j < n; ++j) {
                if (isConnected[i][j])
                    uf.add(i, j);
            }
        }
        unordered_set<int> roots;
        for (int i = 0; i < n; ++i) {
            // cout << "i = " << i << ", p[i] = " << uf.getParent(i) << endl;
            roots.emplace(uf.getParent(i));

        }
        return roots.size();
    }
};

int main() {
    Solution sol;
    auto con = vector<vector<int>>({{1,0,0}, {0,1,0}, {0,0,1}});
    int s = sol.findCircleNum(con);
    cout << s << endl;
    return 0;
}